### ISSUE TITLE
DL Semantics and Quality of OWLObjectAllValuesFrom

### ISSUE THREAD

**Issue by AjayUmakanth** [dice-group/Ontolearn#339]
I have created a custom owl graph as shown below

```
<?xml version="1.0"?>
<rdf:RDF xmlns="http://www.w3.org/2002/07/owl#"
     xml:base="http://www.w3.org/2002/07/owl"
     xmlns:ns1="http://example.org/"
     xmlns:owl="http://www.w3.org/2002/07/owl#"
     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
     xmlns:xml="http://www.w3.org/XML/1998/namespace"
     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#">
    <Ontology/>
    


    <!-- 
    ///////////////////////////////////////////////////////////////////////////////////////
    //
    // Object Properties
    //
    ///////////////////////////////////////////////////////////////////////////////////////
     -->

    


    <!-- http://example.org/writes -->

    <ObjectProperty rdf:about="http://example.org/writes">
        <rdfs:domain rdf:resource="http://example.org/author"/>
        <rdfs:range rdf:resource="http://example.org/paper"/>
    </ObjectProperty>
    


    <!-- http://example.org/written_by -->

    <ObjectProperty rdf:about="http://example.org/written_by">
        <rdfs:domain rdf:resource="http://example.org/paper"/>
        <rdfs:range rdf:resource="http://example.org/author"/>
    </ObjectProperty>
    


    <!-- 
    ///////////////////////////////////////////////////////////////////////////////////////
    //
    // Data properties
    //
    ///////////////////////////////////////////////////////////////////////////////////////
     -->

    


    <!-- http://example.org/author_property_1 -->

    <DatatypeProperty rdf:about="http://example.org/author_property_1">
        <rdfs:domain rdf:resource="http://example.org/author"/>
        <rdfs:range rdf:resource="http://www.w3.org/2001/XMLSchema#double"/>
    </DatatypeProperty>
    


    <!-- http://example.org/author_property_2 -->

    <DatatypeProperty rdf:about="http://example.org/author_property_2">
        <rdfs:domain rdf:resource="http://example.org/author"/>
        <rdfs:range rdf:resource="http://www.w3.org/2001/XMLSchema#double"/>
    </DatatypeProperty>
    


    <!-- http://example.org/paper_property_1 -->

    <DatatypeProperty rdf:about="http://example.org/paper_property_1">
        <rdfs:domain rdf:resource="http://example.org/paper"/>
        <rdfs:range rdf:resource="http://www.w3.org/2001/XMLSchema#double"/>
    </DatatypeProperty>
    


    <!-- http://example.org/paper_property_2 -->

    <DatatypeProperty rdf:about="http://example.org/paper_property_2">
        <rdfs:domain rdf:resource="http://example.org/paper"/>
        <rdfs:range rdf:resource="http://www.w3.org/2001/XMLSchema#double"/>
    </DatatypeProperty>
    


    <!-- 
    ///////////////////////////////////////////////////////////////////////////////////////
    //
    // Classes
    //
    ///////////////////////////////////////////////////////////////////////////////////////
     -->

    


    <!-- http://example.org/author -->

    <Class rdf:about="http://example.org/author"/>
    


    <!-- http://example.org/paper -->

    <Class rdf:about="http://example.org/paper"/>
    


    <!-- 
    ///////////////////////////////////////////////////////////////////////////////////////
    //
    // Individuals
    //
    ///////////////////////////////////////////////////////////////////////////////////////
     -->

    


    <!-- http://example.org/author#1 -->

    <NamedIndividual rdf:about="http://example.org/author#1">
        <rdf:type rdf:resource="http://example.org/author"/>
        <ns1:writes rdf:resource="http://example.org/paper#1"/>
        <ns1:writes rdf:resource="http://example.org/paper#2"/>
        <ns1:author_property_2 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">1.0</ns1:author_property_2>
    </NamedIndividual>
    


    <!-- http://example.org/author#2 -->

    <NamedIndividual rdf:about="http://example.org/author#2">
        <rdf:type rdf:resource="http://example.org/author"/>
        <ns1:writes rdf:resource="http://example.org/paper#2"/>
        <ns1:author_property_1 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">0.6000000238418579</ns1:author_property_1>
        <ns1:author_property_2 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">0.30000001192092896</ns1:author_property_2>
    </NamedIndividual>
    


    <!-- http://example.org/author#3 -->

    <NamedIndividual rdf:about="http://example.org/author#3">
        <rdf:type rdf:resource="http://example.org/author"/>
        <ns1:writes rdf:resource="http://example.org/paper#3"/>
        <ns1:writes rdf:resource="http://example.org/paper#4"/>
        <ns1:author_property_1 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">0.10000000149011612</ns1:author_property_1>
        <ns1:author_property_2 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">1.0</ns1:author_property_2>
    </NamedIndividual>
    


    <!-- http://example.org/author#4 -->

    <NamedIndividual rdf:about="http://example.org/author#4">
        <rdf:type rdf:resource="http://example.org/author"/>
        <ns1:writes rdf:resource="http://example.org/paper#4"/>
        <ns1:author_property_2 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">1.0</ns1:author_property_2>
    </NamedIndividual>
    


    <!-- http://example.org/paper#1 -->

    <NamedIndividual rdf:about="http://example.org/paper#1">
        <rdf:type rdf:resource="http://example.org/paper"/>
        <ns1:written_by rdf:resource="http://example.org/author#1"/>
        <ns1:paper_property_1 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">0.8999999761581421</ns1:paper_property_1>
        <ns1:paper_property_2 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">0.800000011920929</ns1:paper_property_2>
    </NamedIndividual>
    


    <!-- http://example.org/paper#2 -->

    <NamedIndividual rdf:about="http://example.org/paper#2">
        <rdf:type rdf:resource="http://example.org/paper"/>
        <ns1:written_by rdf:resource="http://example.org/author#1"/>
        <ns1:written_by rdf:resource="http://example.org/author#2"/>
        <ns1:paper_property_1 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">0.10000000149011612</ns1:paper_property_1>
        <ns1:paper_property_2 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">0.20000000298023224</ns1:paper_property_2>
    </NamedIndividual>
    


    <!-- http://example.org/paper#3 -->

    <NamedIndividual rdf:about="http://example.org/paper#3">
        <rdf:type rdf:resource="http://example.org/paper"/>
        <ns1:written_by rdf:resource="http://example.org/author#3"/>
        <ns1:paper_property_1 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">0.699999988079071</ns1:paper_property_1>
        <ns1:paper_property_2 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">0.949999988079071</ns1:paper_property_2>
    </NamedIndividual>
    


    <!-- http://example.org/paper#4 -->

    <NamedIndividual rdf:about="http://example.org/paper#4">
        <rdf:type rdf:resource="http://example.org/paper"/>
        <ns1:written_by rdf:resource="http://example.org/author#3"/>
        <ns1:written_by rdf:resource="http://example.org/author#4"/>
        <ns1:paper_property_1 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">0.10000000149011612</ns1:paper_property_1>
        <ns1:paper_property_2 rdf:datatype="http://www.w3.org/2001/XMLSchema#double">0.30000001192092896</ns1:paper_property_2>
    </NamedIndividual>
</rdf:RDF>

```

This is my code

```
from ontolearn.knowledge_base import KnowledgeBase
from ontolearn.concept_learner import EvoLearner
from ontolearn.learning_problem import PosNegLPStandard
from ontolearn.owlapy.model import OWLNamedIndividual, IRI

knowledgeBase = KnowledgeBase(path="./customGraph.owl")  #this is the path pointing to the above graph
model = EvoLearner(knowledge_base=knowledgeBase, use_data_properties=False)

positive_examples = ["http://example.org/author#1","http://example.org/author#3"]
negative_examples = ["http://example.org/author#2","http://example.org/author#4","http://example.org/paper#1","http://example.org/paper#2","http://example.org/paper#3","http://example.org/paper#4"]

typed_pos = set(map(OWLNamedIndividual, map(IRI.create, set(positive_examples))))
typed_neg = set(map(OWLNamedIndividual, map(IRI.create, set(negative_examples))))
lp = PosNegLPStandard(pos=typed_pos, neg=typed_neg)
model.fit(lp)
hypotheses = list(model.best_hypotheses(n=10))

[print(_) for _ in hypotheses]
```

When I run the code above, this is what I get
```
<class 'ontolearn.search.EvoLearnerNode'> at 0xa6c1fd	(≥ 2 writes.paper) ⊓ ⊤	Quality:1.0	Length:6	Tree Length:5	Tree Depth:2	|Indv.|:2
<class 'ontolearn.search.EvoLearnerNode'> at 0xa6c1f3	∀ written_by.paper	Quality:0.75	Length:3	Tree Length:2	Tree Depth:1	|Indv.|:4
<class 'ontolearn.search.EvoLearnerNode'> at 0xa6c1da	¬paper	Quality:0.75	Length:2	Tree Length:2	Tree Depth:1	|Indv.|:4
<class 'ontolearn.search.EvoLearnerNode'> at 0xa6c1c6	author ⊓ ⊤	Quality:0.75	Length:3	Tree Length:3	Tree Depth:1	|Indv.|:4
<class 'ontolearn.search.EvoLearnerNode'> at 0xa6c1cb	author ⊔ author	Quality:0.75	Length:3	Tree Length:3	Tree Depth:1	|Indv.|:4
<class 'ontolearn.search.EvoLearnerNode'> at 0xa6c1e4	author ⊓ author	Quality:0.75	Length:3	Tree Length:3	Tree Depth:1	|Indv.|:4
<class 'ontolearn.search.EvoLearnerNode'> at 0xa6c1a8	⊤ ⊓ author	Quality:0.75	Length:3	Tree Length:3	Tree Depth:1	|Indv.|:4
```

In the above result you can see that in the second hypothesis '∀ written_by.paper' we get a quality function (accuracy by default) of 0.75. But, in the graph the domain and range of the object property written_by is paper and author respectively. So, the hypothesis  structure itself is wrong and the quality function also is wrong as it would be 0 (as there would be no class written by a paper)

These kinds of results **do not occur always**, but does occue a considerable amout of times (approximately 20% of the times when I tried). This is a small custom graph, but it occurs on a real world big graph too and is many times the top hypothesis which returns an incorrect quality value.

Let me know if I am doing anything wrong, or is there an issue in evolearner

**Comment by Demirrr** [comment]
Dear @AjayUmakanth,

Thank you for your interest and your question. 

The below code snippet show that ∀ written_by.paper has 0.75 Acc given your positive and negative examples.

Why do you think that the accuracy of ∀ written_by.paper should be 0 ?

```python
from ontolearn.knowledge_base import KnowledgeBase
from owlapy.model import IRI, OWLClass, OWLObjectSomeValuesFrom, OWLObjectAllValuesFrom
from owlapy.model import OWLClass, OWLNamedIndividual
from owlapy.render import DLSyntaxObjectRenderer

def acc(*, individuals, pos, neg):
    tp = len(pos.intersection(individuals))
    tn = len(neg.difference(individuals))

    fp = len(neg.intersection(individuals))
    fn = len(pos.difference(individuals))

    return (tp+tn)/(tp+tn+fp+fn)

# loading
knowledgeBase = KnowledgeBase(path="customGraph.owl")
dl_renderer = DLSyntaxObjectRenderer()
positive_examples = ["http://example.org/author#1", "http://example.org/author#3"]
negative_examples = ["http://example.org/author#2", "http://example.org/author#4",
                     "http://example.org/paper#1", "http://example.org/paper#2", "http://example.org/paper#3",
                     "http://example.org/paper#4"]
typed_pos = {OWLNamedIndividual(IRI.create(i)) for i in positive_examples}
typed_neg = {OWLNamedIndividual(IRI.create(i)) for i in negative_examples}
for i in knowledgeBase.concepts:
    print(f"{dl_renderer.render(i)}\t Acc:{acc(individuals={ind for ind in knowledgeBase.individuals(i)}, pos=typed_pos, neg=typed_neg)}")
    for j in knowledgeBase.object_properties:
        c = OWLObjectSomeValuesFrom(property=j, filler=i)
        print(f"{dl_renderer.render(c)}\t Acc:{acc(individuals={ind for ind in knowledgeBase.individuals(c)}, pos=typed_pos, neg=typed_neg)}")
        c = OWLObjectAllValuesFrom(property=j, filler=i)
        print(f"{dl_renderer.render(c)}\t Acc:{acc(individuals={ind for ind in knowledgeBase.individuals(c)}, pos=typed_pos, neg=typed_neg)}")

"""
paper	 Acc:0.25
∃ written_by.paper	 Acc:0.75
∀ written_by.paper	 Acc:0.75
∃ writes.paper	 Acc:0.75
∀ writes.paper	 Acc:0.25
author	 Acc:0.75
∃ written_by.author	 Acc:0.25
∀ written_by.author	 Acc:0.25
∃ writes.author	 Acc:0.75
∀ writes.author	 Acc:0.25
"""
```
